orttraining/orttraining/lazy_tensor/accelerator.cpp

```diff
@@ -0,0 +1,488 @@
+  NvtxRange range(__func__);
+#endif
+  TORCH_CHECK(inputs.size(), "Need at least one input.");
+  for (const auto& input : inputs) {
+    TORCH_CHECK(input.isTensor() || input.isScalar(), "Compiler can only handle Tensor or Scalar inputs.");
+  }
+}
+
+// Store input types in sub-graph so that
+// ONNX exporter can use them. Input types
+// are required when executing ONNX model
+// in ORT.
+// TODO: Allow ORT to accept models without
+// input types. Then, we can remove this function.
+static void PropagateArgTypes(
+    const at::ArrayRef<c10::IValue>& inputs,
+    std::shared_ptr<torch::jit::Graph> graph) {
+  TORCH_CHECK(graph->inputs().size() == inputs.size(),
+              "Number of provided inputs must match captured sub-graph's schema.");
+  for (size_t i = 0; i < graph->inputs().size(); ++i) {
+    auto input_symbol = graph->inputs()[i];
+    auto input_value = inputs[i];
+    if (!input_value.isTensor()) {
+      // The allowed IR components in ONNX exporter and Pytorch
+      // are a little different. I am not confident to fill
+      // types other than tensor, because of the ambiguous scalar
+      // representations in Pytorch.
+      continue;
+    }
+    input_symbol->setType(input_value.type());
+  }
+}
+
+// ONNX exporter is written in Python, so
+// this function may call some Python functions.
+// Be aware of GIL issue.
+// The returned value is the path to exported
+// ONNX file.
+static std::string ExportToOnnx(
+    std::shared_ptr<torch::jit::Graph> graph,
+    const at::ArrayRef<c10::IValue>& args) {
+#ifdef USE_CUDA
+  NvtxRange range(__func__);
+#endif
+  // ONNX exporter modifies the graph in-place, so we
+  // need to clone it to avoid interaction between
+  // Pytorch's JIT mechanism and ONNX graph.
+  std::shared_ptr<torch::jit::Graph> new_subgraph(graph->copyUnique().release());
+  // Acquire GIL since Python is not multi-threading.
+  pybind11::gil_scoped_acquire guard{};
+  // Retrieve Python exporter function.
+  pybind11::function export_to_onnx =
+      pybind11::reinterpret_borrow<pybind11::function>(
+          pybind11::module::import("onnxruntime.training.experimental.exporter").attr("_export_jit_graph_to_onnx_model_proto"));
+  // Fill types up. The sub-graphp from LazyTensor doesn't
+  // contain input shapes.
+  PropagateArgTypes(args, new_subgraph);
+  // Execute Python function.
+  auto result = export_to_onnx(new_subgraph, ::torch::onnx::OperatorExportTypes::ONNX);
+  return result.cast<std::string>();
+}
+
+// Create an empty session object.
+// Models will be loaded later.
+static std::unique_ptr<onnxruntime::InferenceSession> CreateSession() {
+#ifdef USE_CUDA
+  NvtxRange range(__func__);
+#endif
+  // Enviroment shared by all sessions.
+  static onnxruntime::Environment& pybind_default_env = GetLtcEnv();
+  // All sessions use the same config.
+  static onnxruntime::SessionOptions sess_opts;
+  return std::make_unique<onnxruntime::InferenceSession>(sess_opts, pybind_default_env);
```

Comment:
where is the Cuda execution provider get registered?

Reply:
In `class CUDAExecutionProviderPool`, it creates for one EP for each GPU.